src/lib.rs

```diff
@@ -0,0 +1,1036 @@
+    }
+
+    /// Determine whether an instruction is a move; if so, return the
+    /// Operands for (src, dst).
+    fn is_move(&self, insn: Inst) -> Option<(Operand, Operand)>;
+
+    // --------------------------
+    // Instruction register slots
+    // --------------------------
+
+    /// Get the Operands for an instruction.
+    fn inst_operands(&self, insn: Inst) -> &[Operand];
+
+    /// Get the clobbers for an instruction.
+    fn inst_clobbers(&self, insn: Inst) -> &[PReg];
+
+    /// Get the number of `VReg` in use in this function.
+    fn num_vregs(&self) -> usize;
+
+    /// Get the VRegs that are pointer/reference types. This has the
+    /// following effects for each such vreg:
+    ///
+    /// - At all safepoint instructions, the vreg will be in a
+    ///   SpillSlot, not in a register.
+    /// - The vreg *may not* be used as a register operand on
+    ///   safepoint instructions: this is because a vreg can only live
+    ///   in one place at a time. The client should copy the value to an
+    ///   integer-typed vreg and use this to pass a pointer as an input
+    ///   to a safepoint instruction (such as a function call).
+    /// - At all safepoint instructions, all live vregs' locations
+    ///   will be included in a list in the `Output` below, so that
+    ///   pointer-inspecting/updating functionality (such as a moving
+    ///   garbage collector) may observe and edit their values.
+    fn reftype_vregs(&self) -> &[VReg] {
+        &[]
+    }
+
+    /// Get the VRegs for which we should generate value-location
+    /// metadata for debugging purposes. This can be used to generate
+    /// e.g. DWARF with valid prgram-point ranges for each value
+    /// expression in a way that is more efficient than a post-hoc
+    /// analysis of the allocator's output.
+    ///
+    /// Each tuple is (vreg, inclusive_start, exclusive_end,
+    /// label). In the `Output` there will be (label, inclusive_start,
+    /// exclusive_end, alloc)` tuples. The ranges may not exactly
+    /// match -- specifically, the returned metadata may cover only a
+    /// subset of the requested ranges -- if the value is not live for
+    /// the entire requested ranges.
+    fn debug_value_labels(&self) -> &[(Inst, Inst, VReg, u32)] {
+        &[]
+    }
+
+    /// Is the given vreg pinned to a preg? If so, every use of the
+    /// vreg is automatically assigned to the preg, and live-ranges of
+    /// the vreg allocate the preg exclusively (are not spilled
+    /// elsewhere). The user must take care not to have too many live
+    /// pinned vregs such that allocation is no longer possible;
+    /// liverange computation will check that this is the case (that
+    /// there are enough remaining allocatable pregs of every class to
+    /// hold all Reg-constrained operands).
+    fn is_pinned_vreg(&self, _: VReg) -> Option<PReg> {
+        None
+    }
+
+    /// Return a list of all pinned vregs.
+    fn pinned_vregs(&self) -> &[VReg] {
```

Comment:
This function isn't actually needed at all: `is_pinned_vreg` alone is sufficient and can be used to initialize `is_pinned` in vregs.

Reply:
`is_pinned_vreg` is *technically* sufficient, but using that alone is really inefficient: consider a function with thousands of vregs and just 32..64 or so are pinned. This was the reason I added this function as well.

Reply:
Consider the implementation in your regalloc.rs branch which wraps regalloc2:

```rust
    fn is_pinned_vreg(&self, vreg: regalloc2::VReg) -> Option<regalloc2::PReg> {
        if vreg.vreg() < self.ra2_env().vreg_offset {
            Some(regalloc2::PReg::new(vreg.vreg(), vreg.class()))
        } else {
            None
        }
    }
```

`is_pinned_vreg(vreg).is_some()` has practically zero overhead after inlining in the loop that initializes the vregs. I would expect that most clients of regalloc2 end up pinning the first few vregs so the implementation should always be similar.

However creating a `Vec` for the pinned regs isn't *too* much of a hassle, so I don't mind that much.

Reply:
Yes, that's fair: I'd hope that in release mode, rustc would be able to see that the loop terminates early. However, I don't necessarily like relying on such an optimization to make the API efficient, and I don't like having hidden conditions or cliffs (such as "put all the vregs at the beginning") that significantly affect efficiency. Or rather, if we have such a condition, we should put it in the API ("max pinned vreg number" or somesuch).

I'd be happy to go with "max pinned vreg number" instead if you'd prefer that -- it sounds like it would work for both my use-case and yours.